utils/common/src/main/java/org/apache/brooklyn/util/stream/LoggingOutputStream.java

```diff
@@ -0,0 +1,134 @@
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.brooklyn.util.stream;
+
+import java.io.FilterOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.slf4j.Logger;
+
+/**
+ * Wraps another output stream, intercepting the writes to log it.
+ */
+public class LoggingOutputStream extends FilterOutputStream {
+
+    private static final OutputStream NOOP_OUTPUT_STREAM = new FilterOutputStream(null) {
+        @Override public void write(int b) throws IOException {
+        }
+        @Override public void flush() throws IOException {
+        }
+        @Override public void close() throws IOException {
+        }        
+    };
+    
+    public static Builder builder() {
+        return new Builder();
+    }
+    
+    public static class Builder {
+        OutputStream out;
+        Logger log;
+        String logPrefix;
+        
+        public Builder outputStream(OutputStream val) {
+            this.out = val;
+            return this;
+        }
+        public Builder logger(Logger val) {
+            this.log = val;
+            return this;
+        }
+        public Builder logPrefix(String val) {
+            this.logPrefix = val;
+            return this;
+        }
+        public LoggingOutputStream build() {
+            return new LoggingOutputStream(this);
+        }
+    }
+    
+    protected final Logger log;
+    protected final String logPrefix;
+    private final AtomicBoolean running = new AtomicBoolean(true);
+    private final StringBuilder lineSoFar = new StringBuilder(16);
+
+    private LoggingOutputStream(Builder builder) {
+        super(builder.out != null ? builder.out : NOOP_OUTPUT_STREAM);
+        log = builder.log;
+        logPrefix = (builder.logPrefix != null) ? builder.logPrefix : "";
+      }
+
+    @Override
+    public void write(int b) throws IOException {
+        if (running.get() && b >= 0) onChar(b);
+        out.write(b);
+    }
+
+    @Override
+    public void flush() throws IOException {
+        try {
+            if (lineSoFar.length() > 0) {
+                onLine(lineSoFar.toString());
+                lineSoFar.setLength(0);
```

Comment:
Is your concern about leaking that the `lineSoFar` might now have a very large internal array in `StringBuilder.value`, which won't be GC'ed until the `LoggingOutputStream` is GC'ed?

I could add a call to `StringBuilder.trimToSize()` as well (which would set it to size 0, rather than our initial size of 16). That would be less efficient (causing array-copy to grow again), but probably not noticeably so. Maybe I'll compromise and say that if the size was `>= 1024` then I'll `trimToSize`.

Note that this code was copy-pasted from `StreamGobbler`.